src/Features/Protocol/LanguageServices/RoslynLanguageService.cs

```diff
@@ -0,0 +1,673 @@
+
+                while (node != null)
+                {
+                    var symbol = model.GetDeclaredSymbol(node);
+                    if (symbol != null)
+                    {
+                        return symbol;
+                    }
+
+                    node = node.Parent;
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Answers a workspace symbols request by providing a list of symbols found in a given workspace.
+        /// https://microsoft.github.io/language-server-protocol/specification#workspace_symbol
+        /// </summary>
+        /// <param name="solution">the current solution.</param>
+        /// <param name="request">the workspace request with the query to invoke.</param>
+        /// <param name="cancellationToken">a cancellation token.</param>
+        /// <returns>a list of symbols in the workspace.</returns>
+        public async Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(Solution solution, WorkspaceSymbolParams request, CancellationToken cancellationToken)
+        {
+            var symbols = ArrayBuilder<SymbolInformation>.GetInstance();
+
+            if (solution == null)
+            {
+                return symbols.ToArrayAndFree();
+            }
+
+            var searchTasks = solution.Projects.Select(
+                p => Task.Run(() => SearchProjectAsync(p), cancellationToken)).ToArray();
+
+            await Task.WhenAll(searchTasks).ConfigureAwait(false);
+
+            return symbols.ToArrayAndFree();
+
+            // local functions
+            async Task SearchProjectAsync(Project project)
+            {
+                var searchService = project.LanguageServices.GetService<INavigateToSearchService_RemoveInterfaceAboveAndRenameThisAfterInternalsVisibleToUsersUpdate>();
+                if (searchService != null)
+                {
+                    // TODO - Update Kinds Provided to return all necessary symbols.
+                    // https://github.com/dotnet/roslyn/projects/45#card-20033822
+                    var items = await searchService.SearchProjectAsync(
+                        project,
+                        ImmutableArray<Document>.Empty,
+                        request.Query,
+                        searchService.KindsProvided,
+                        cancellationToken).ConfigureAwait(false);
+
+                    foreach (var item in items)
+                    {
+                        var symbolInfo = new SymbolInformation
+                        {
+                            Name = item.Name,
+                            Kind = item.Kind.GetKind(),
+                            Location = await item.NavigableItem.Document.ToLocationAsync(item.NavigableItem.SourceSpan, cancellationToken).ConfigureAwait(false),
+                        };
+
+                        lock (symbols)
+                        {
+                            symbols.Add(symbolInfo);
```

Comment:
high contention.  instead of doing this, just have each project search return a list of results, after your 'whenall' above, just merge them.  no contention at all there.